src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +319,99 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeoutDelay q = do
+  timedOutVar ← registerDelay timeoutDelay
+
+  let
+    readItems xs =
+      readTBMQueue q ≫= \case
+        Nothing → return xs
+        Just x → go (x:xs)
+
+    timeout =
+      readTVar timedOutVar ≫= check
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = readItems xs <|> xs <$ timeout
+
+  atomically $ go []
```

Comment:
I played around with this code, and the following seems to happen: 

1. I put `1000` items in the queue, and I try to bite off `2000` items in `5s`.
2. `go` recurses until it has got `1000` items, and the timeout has not yet fired; then because `not t` is true, then `readItems` is called which blocks until another item is added to the queue. Meanwhile, the timeout fires, but nobody hears about it and we just block.